### ISSUE TITLE
FIDO misidentifies dummy plain text file

### ISSUE THREAD

**Issue by UmbrellaDish** [openpreserve/fido#114]
## Dev Effort
0.5D

## Description
I have a simple text file containing following lines:

        Many examples are known to be questionable in significance.
        You might find this is an example of those.

The output of FIDO is:

FIDO v1.3.6 (formats-v92.xml, container-signature-20170330.xml, format_extensions.xml)
OK,117,x-fmt/111,"Plain Text File","External",102,"/tmp/example.txt","text/plain","extension"
OK,117,fmt/1085,"TRIM Context Reference File","External",102,"/tmp/example.txt","None","extension"
FIDO: Processed      1 files in 175.41 msec,  6 files/sec

Second line is unexpected: As the file has the extension .txt but does not begin with "[TRIM Context Reference]" as suggested in PRONOM database, seen in signatures tab of the webinterface, I wonder why FIDO refuse looking closer at it. I have had it download the current signatures yesterday.

**Comment by jrwdunham** [comment]
@UmbrellaDish I believe you are seeing the expected behaviour. See http://wiki.opf-labs.org/display/KB/FIDO+usage+guide#FIDOusageguide-Analysingasinglefile and the "Matchtypes" section. FIDO is unable to match your test file based on any signature so it can only identify the file based on its extension. Since fmt/1085 and x-fmt/111 are the only PRONOM formats with the .txt extension, they are both returned. Search for `<extension>txt</extension>` in fido/conf/formats-v92.xml.

If I am missing something and you do not agree that this is the expected behaviour, please explain why. Thanks.

**Comment by UmbrellaDish** [comment]
In my eyes, if there are multiple criteria specified for a given PRONOM id, all of them *MUST* be matched and ensured matching. So, if it is specified that a file type has a certain extention plus a certain preamble and the file type is returned (albeit not as the first one, just luck?) notwithstanding the lack of the preamble, then I am wondering what actually triggers FIDO to evaluate the signature/pattern/regex=`(?s)\A\[TRIM Context Reference\]?`.

As we use FIDO in an automatic background process, DROID user interface cannot be of much help, I am afraid. It requires even some workarounds embedding it in a python module, as it seems to insist on being called from the command line.

**Comment by jrwdunham** [comment]
@UmbrellaDish FIDO's algorithm (simplified) is to first find all signature-based matches and if it finds none then return all extension-based matches. Maybe a solution here would be to implement a flag that would allow the user to tell FIDO to not perform extension-based matches at all. I think clearly we want a file with the wrong extension to still be identified as a particular format if it matches that format's signature.

**Comment by UmbrellaDish** [comment]
That algorithm may work for many file types pretty well and it is the way to go in order not to rely too strictly on filename extensions. But if the signature matches, the extension does not need to be checked. So I conclude that there is no need to add a flag, or to fiddle with the inner workings, just to correct the file type definition. Delete ".txt" extension restriction. If the preamble is present, the file extension may well be ".txt", but as you say how the algorithm is implemented it does not matter indeed.

In my humble opinion, however, you should check the extensions at any rate. In case the TRIM-line is present, but the extension is ".html" or ".zip" or whatsoever, mere reason suggests an error condition. At last, you need to distinguish general extensions (as .txt, .xml, ...) characterized by the requirement that the signature must also match, and specific extensions that define the type all on their own.

**Comment by jrwdunham** [comment]
Let's be clear about what is happening in this case.

1. The x-fmt/111 format has no signature in the DROID signature file. All it has is the associated extension `txt`.
2. The fmt/1085 format does have a signature in the DROID signature file; FIDO translates that signature to this Python regex: `(?s)\A\[TRIM Context Reference\]`. The fmt/1085 format also has two associated extensions: `txt` and `tr5`.

When FIDO attempts to identify a file with extension `txt` that contains the string provided in the first comment of this issue, it is unable to find a match based on PRONOM signatures. This is because x-fmt/111 has no signature and because the fmt/1085 signature (now a Python regex) does not match the bytes of the file.

Therefore FIDO finds all PRONOM formats with an extension that matches `txt`, it finds x-fmt/111 and fmt/1085, and it prints both to stdout.

Given the above, I am confused when you say

> But if the signature matches, the extension does not need to be checked.

because no signatures match in this case.

I have tested the same file with the [Siegfried](https://github.com/richardlehane/siegfried) tool and it only returns x-fmt/111, also on the basis of an extension match: `basis   : 'extension match txt; text match ASCII'`. Messing around with the siegfried code, I can see that it too matches fmt/1085 to this file at some intermediary stage. However, it appears to arbitrarily choose the first extension-based match and prints only that match, i.e., x-fmt/111. If the choice is in fact arbitrary here, it is not clear to me that the Siegfried strategy is preferable.

If the goal is to disassociate the `txt` extension from the fmt/1085 format, then this is not an issue for FIDO but a matter that would best be fixed by submitting a request for modification to [PRONOM](https://www.nationalarchives.gov.uk/PRONOM/Default.aspx) itself.

Please understand that I am doing my best to understand and resolve the issue at hand. If I am making false statements or failing to understand something, please let me know.

**Comment by richardlehane** [comment]
Just to add a little more detail re. siegfried's approach in these cases, the relevant bit of code is: https://github.com/richardlehane/siegfried/blob/master/pkg/pronom/identifier.go#L234.

A few things:

  1) sf isn't only matching the extension for this file, it has a text match too (the `text match ASCII` bit in the basis field) that has added strength to the match. In fact, even if you changed or removed the extension it would likely still match x-fmt/111 on the basis of a text match alone. This e.g. allows sf to match README files etc. as x-fmt/111
  2) assuming this wasn't actually a text file but a binary file with a ".txt" extension (sometimes happens). You'd get an UNKNOWN result from sf. This is because sf actively **excludes** file extension matches where it expects other type of matching to happen (i.e. if it expects a text match in the case of x-fmt/111, or a binary signature match in the case of fmt/1085, or in other cases where it might expect a container signature match - then it will exclude those from the list of possibilities).
  3) finally, ignoring the whole text thing altogether - if this was some other kind of file where multiple file extension matches have occurred, but without any other stronger kind of identification, and those file extension matches can't be excluded because there aren't any strong signatures for those file types - in those cases sf will also report UNKNOWN and include the list of file extension matches in a warning message saying it may be one of those. I.e. sf will only report PUIDs based on a file extension match alone if there is only a single viable candidate. The rationale is that this isn't a strong basis for a match and if there are other possibilities best to just report unknown and leave it to the user to work out based on the warning message.

So that's the way siegfried handles these cases but I should be clear it doesn't mean I think FIDO is wrong. At the end of the day, both approaches feel valid to me and technically "correct". It is more a matter of taste. Do you lean towards providing as much info as possible to users or do you favour more certainty in the results? Agree with @jrwdunham that you'd want to be careful about changing FIDO's approach here as your users will be already conditioned to expect the current approach.

**Comment by UmbrellaDish** [comment]
Am 19.04.2018 um 20:57 schrieb Joel Dunham:
> However, it appears to arbitrarily choose the first extension-based 
> match and prints only that match

In my opinion the choice is not arbitrary, or rather it should not be. 
The logic applied is a little bit too lax. It is unfortunate if mere 
signature mismatch makes it conclude that the file can be one of a 
couple of specific types based on file extension and so to just pick any 
by random is really far from an option.

.xml, .txt, .yaml, .conf etc. are file extensions that characterize the 
mere technical data format, whatever their purpose on application level, 
they are hints regarding how the data is to be read, what generic 
parsers are applicable. Whereas .html, .svg, .png, .flac etc. describe 
also the application class: .html can be viewed in a browser, .flac can 
be made audible, etc.

This distinction format- vs. application-related file type is very 
relevant. In conclusion FIDO I guess needs to check:

     ext .tr5 OR (ext .txt AND regex (?s)\A\[TRIM Context Reference\])

or, more general, any_of(extensions.APPLICATION_CLASS) OR 
(any_of(extensions.FORMAT_CLASS) AND SIGNATURE_MATCH). However, possible 
format extensions and signatures are correlated. They exclude each other 
in the way that e.g. a signature checking XML-tags cannot apply combined 
with .yaml.

In other words, a proper file type detector should not check format 
class exceptions without reconsidering carefully if any content-based 
signatures match, too. If none match, the tool must yield only the most 
appropriate generic description. That way, it avoids these pitfalls 
accidentally laid out by others who were just not careful or lucky 
enough in choosing a good file type extension related to their tools.

Given FIDO and how it detects the file type, properly or not, we can 
only resort to picking its first guess. I hope it will be right in most 
cases.



-- 
UB Heidelberg (Altstadt)
Plöck 107-109, 69117 HD
Abt. Informationstechnik
Tel. 06221 / 54 3550
http://www.ub.uni-heidelberg.de/

**Comment by ablwr** [comment]
This seems to have been resolved sometime over a year ago. After the new release, if this is still a concern, please open a new issue.
